docs/design/operator.md

```diff
@@ -0,0 +1,387 @@
+deployment, configuration, and management of these drivers using new Kubernetes
+APIs defined as a set of Custom Resource Definitions (CRDs).
+
+## System Overview
+
+The Ceph-CSI Operator is a Kubernetes operator that offers a Kubernetes native
+management interfaces for Ceph CSI drivers. It provides a set of new Kubernetes
+APIs (CRDs) for governing deployment and management of Ceph-CSI drivers, as
+well as connection details to Ceph clusters. The operator ensures that the
+desired state specified in the CRDs is maintained in the cluster.
+
+## Assumptions and Dependencies
+
+Kubernetes cluster is available and operational. Ceph cluster is set up and
+accessible from the Kubernetes cluster. Users are familiar with Kubernetes and
+Ceph concepts. CRD and operator versions are compatible with the Kubernetes
+version in use.
+
+## Goals and Guidelines
+
+- Automate the deployment and lifecycle for Ceph-CSI drivers.
+- Provide a consistent and flexible configuration mechanism using CRDs.
+- Enable namespace-scoped configurations for different Ceph-CSI drivers.
+- Ensure high availability and scalability of the operator.
+- Support multiple deployment modes for the operator:
+  - Own-namespace: the operator manages a single Ceph-CSI driver in the current
+    namespace where operator is deployed.
+  - Multi-namespace: the operator manages Ceph-CSI drivers in user-specified
+    namespaces.
+  - All-namespaces: the operator manages Ceph-CSI drivers in all namespaces.
+
+## Development Methods
+
+Use kubebuilder framework which is highly used in operator development,
+following Kubernetes best practices for controllers and operators
+
+## Architecture
+
+```mermaid
+graph TD
+    A[CRD Changes] --> B[Operator]
+    B -->  D[Configure CephFS CSI Driver]
+    B -->  E[Configure RBD CSI Driver]
+    B -->  F[Configure NFS CSI Driver]
+
+```
+
+In this diagram:
+
+- **CRD Changes**: Represents changes made to Custom Resource Definitions,
+  which trigger actions in the operator.
+- **Operator**: Listens for changes in CRDs and initiates the installation and
+  configuration of CSI drivers.
+- **Configure CephFS, NFS, RBD**: Actions performed by the operator to install
+  and configure the respective CSI drivers based on the CRD changes.
+
+## CRDs for ceph-csi-operator
+
+### CephCSIOperatorConfig CRD
+
+Manages operator-level configurations and offers a place to overwrite settings
+for CSI drivers. This CRD is a namespace-scoped CRD and a single CR named
+ceph-csi-operator-config can be created by an admin inside the operator's
+namespace.
+
+The configurations are catagorized into 2 different types
+
+- Operator configurations
+- Default configuration overwrites applied to all Ceph CSI drivers
+
+```yaml
+---
+kind: CephCSIOperatorConfig
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: ceph-csi-operator-config
+  namespace: <operator-namespace>
+spec:
+  logLevel: 1
+  driverSpecDefaults:
```

Comment:
IMO, the deployment mode (the operator's 'watch' scope) control makes the most sense here


something like...
```yaml
spec:
  watchNamespaces: # if empty, only watch current ns, ['*'] means watch all
    - ns1
    - ns2
    - etc
```

Reply:
sounds good, am opening a new issue for it. https://github.com/ceph/ceph-csi-operator/issues/7

Reply:
No, we cannot do that. The deployment control is part of the OLM spec and needs to be identified there in order for this to work. I Don't want us to create an operator that has internally code to reconcile OLM entities.